src/psyclone/tests/domain/gocean/goloop_test.py

```diff
@@ -0,0 +1,164 @@
+        goiloop.gen_code(None)
+
+
+def test_goloop_unmatched_offsets():
+    ''' Attempt to generate code for a loop with constant bounds with
+    two different index offsets '''
+    gosched = GOInvokeSchedule('name', [])
+    gojloop = GOLoop(parent=gosched, loop_type="outer")
+    goiloop = GOLoop(parent=gosched, loop_type="inner")
+    gosched.addchild(gojloop)
+    gojloop.loop_body.addchild(goiloop)
+    gokern1 = GOKern()
+    gokern2 = GOKern()
+    # Set the index-offset of this kernel to a value that is not
+    # supported when using constant loop bounds
+    gokern1._index_offset = "go_offset_ne"
+    gokern2._index_offset = "go_offset_sw"
+    goiloop.loop_body.addchild(gokern1)
+    goiloop.loop_body.addchild(gokern2)
+    with pytest.raises(GenerationError) as excinfo:
+        goiloop.gen_code(None)
+    # Note that the kernels do not have a name, so there is a double space
+    assert "All Kernels must expect the same grid offset but kernel  " \
+        "has offset go_offset_sw which does not match go_offset_ne" \
+        in str(excinfo.value)
+
+
+def test_goloop_bounds_invalid_iteration_space():
+    ''' Check that the _upper/lower_bound() methods raise the expected error
+    if the iteration space is not recognised. '''
+    gosched = GOInvokeSchedule('name', [])
+    gojloop = GOLoop(parent=gosched, loop_type="outer")
+    # Have to turn-off constant loop bounds to get to the error condition
+    gosched._const_loop_bounds = False
+    # Set the iteration space to something invalid
+    gojloop._iteration_space = "broken"
+    with pytest.raises(GenerationError) as err:
+        gojloop._upper_bound()
+    assert "Unrecognised iteration space, 'broken'." in str(err.value)
+    with pytest.raises(GenerationError) as err:
+        gojloop._lower_bound()
+    assert "Unrecognised iteration space, 'broken'." in str(err.value)
+
+
+def test_goloop_grid_prop_ref():
+    ''' Tests for the _grid_property_reference() method. '''
+    _, invoke = get_invoke("test11_different_iterates_over_one_invoke.f90",
+                           API, idx=0)
+    schedule = invoke.schedule
+    loop = schedule.walk(GOLoop)[0]
+    # A simple name should result in a new symbol and a suitable reference
+    assert "hello" not in schedule.symbol_table
+    href = loop._grid_property_reference("hello")
+    hsym = schedule.symbol_table.lookup("hello")
+    assert isinstance(hsym, DataSymbol)
+    assert hsym.datatype == INTEGER_TYPE
+    assert isinstance(href, Reference)
+    # A derived-type reference must be in the form of a format string with
+    # "{0}" at the start.
+    with pytest.raises(NotImplementedError) as err:
+        loop._grid_property_reference("wrong%one")
+    assert ("Supplied grid property is a derived-type reference but does "
+            "not begin with '{0}': 'wrong%one'" in str(err.value))
+    gref = loop._grid_property_reference("{0}%grid%xstart")
+    assert isinstance(gref, Reference)
```

Comment:
isinstance(gref, StructureReference)

and maybe also check that gref.parent point to loop?

Reply:
Oh yes, done.